designs/statefulset-disruption.md

```diff
@@ -0,0 +1,891 @@
+
+![kubelet_shutdown_race](https://github.com/AndrewSirenko/karpenter-provider-aws/assets/68304519/9ccb6c94-ef02-40af-84bf-9471827f3303)
+
+Today, the EBS CSI Driver attempts to work around these races by utilizing a [PreStop hook](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/faq.md#what-steps-can-be-taken-to-mitigate-this-issue) that tries to keep the Node Service alive for an additional [terminationGracePeriodSeconds](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/73b43995a70b8fea51655577e4b475613ab400f1/charts/aws-ebs-csi-driver/values.yaml#L373) until all volumes are unmounted. We will explore the shortcomings of this solution later in [problem A alternative solutions](#alternatives-considered).  
+
+*Note: In addition to the Kubelet race, this delay can happen if stateful pods out-live the CSI Node Pod. E.g. Operator has a statefulset that tolerates all taints and has a longer terminationGracePeriod than EBS CSI Driver.*
+
+### Solutions:
+
+We should:
+- A1: Fix Kubelet race condition upstream for future Kubernetes versions
+- A2: Have Karpenter taint terminated nodes as `out-of-service` before removing its finalizer. 
+
+#### A1: Fix race at Kubelet level
+
+The Kubelet Shutdown Manager should not kill CSI Driver Pods before volumes are unmounted. This change must be made at `kubernetes/kubernetes` level. 
+
+Because this solution does not rely on changes in Karpenter, please see [Active PR #125070](https://github.com/kubernetes/kubernetes/pull/125070) for more information on this solution. 
+
+**Pros:** 
+
+- Other cluster autoscalers will not face this race condition. 
+- Reduces pod migration times
+- Reduces risk of data corruption because relevant CSI Drivers will perform unmount operations required  
+
+**Cons:**
+
+- Unavailable until merged in a version of Kubernetes (Likely Kubernetes `v1.31`) (Possibly able to be backported)
+- If Kubelet Shutdown Manager is not able to run 
+
+#### A2: Taint node as `out-of-service` after termination
+
+While this race should be fixed at the Kubelet level long-term, we still need a solution for earlier versions of Kubernetes.
+
+One solution is to mark terminated nodes via `out-of-service` taint. 
+
+In `v1.26`, Kubernetes enabled the [Non-graceful node shutdown handling feature](https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#non-graceful-node-shutdown) by default. This introduces the `node.kubernetes.io/out-of-service` taint, which can be used to mark a node as permanently shut down. See more context in [appendix D2]() 
+
+Once Karpenter confirms that an instance is terminated, adding this taint to the node object will allow the Attach/Detach Controller to treat the volume as not in use, preventing the 6+ minute delay. 
```

Comment:
Oh interesting....so it's an additional signal to the attach/detach controller. That's actually super cool and I get the value in why we'd want to do this